### ISSUE TITLE
Question about graph replace with keras Adam

### ISSUE THREAD

**Issue by posenhuang** [poolio/unrolled_gan#4]
Hi,
Thanks for the nice example. I have a question when you said ``TensorFlow's built-in optimizers use custom C++ code for efficiency, and do not construct a symbolic graph that is differentiable. For this notebook, we use the optimization routines from keras to compute update.`` 
It means that you have to use a keras optimizer to keep the graph differentiable (whereas tensorflow optimizer or ``tf.gradient`` cannot)?
Thanks!

**Comment by poolio** [comment]
You need to have access to a *differentiable* function which takes as input the optimizer state, gradient and old parameters and produces updated parameters and optimizer state:
`new_state, new_params = optimizer_step(cur_state, cur_params, gradient)`
You could easily write these functions from scratch, but Keras' optimization code provides these as well. TensoFlow's built-in optimizers provide a way to update parameters but the `optimizer_step` functoin they use to do so *is not* differentiable. This is because they have combined multiple mathematical operations into one fused C++ op, and have not written a gradient for that op. 

For example, the `MomentumOptimizer` in TensorFlow uses the `apply_momentum` op instead of writing out the mathematical expression for the update (see [here](https://github.com/tensorflow/tensorflow/blob/e64bc92d293e458eb85de7b1f004f9f69b0e247d/tensorflow/python/training/momentum.py#L90) for the `MomentumOptimizer` implementatoin and the `ApplyMomentum` C++ op [here](https://github.com/tensorflow/tensorflow/blob/03beb65cecbc1e49ea477bca7f54543134b31d53/tensorflow/core/kernels/training_ops_gpu.cu.cc#L82).

In contrast, Keras' [implementation](https://github.com/fchollet/keras/blob/2d8739dda9859b91bf2b7da5402d242555c48d7d/keras/optimizers.py#L168) uses existsing TF ops that have gradients defined when implementing momentum, so we are able to backpropagate through the optimization update.
